src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -139,19 +114,30 @@ void HAMUnit<E, J, S, W, C>::ResetParameters() {
 
   parameters = arma::mat(embedCount + searchCount + controllerCount + joinCount + writeCount, 1);
 
-  embed.Parameters() = arma::mat(parameters.memptr(), embedCount, 1, false, true);
-  search.Parameters() = arma::mat(parameters.memptr() + embedCount, searchCount, 1, false, true);
-  controller.Parameters() = arma::mat(parameters.memptr() + embedCount + searchCount, controllerCount, 1, false, true);
-  memory.JoinObject().Parameters() = arma::mat(parameters.memptr() + embedCount + searchCount + controllerCount, joinCount, 1, false, true);
-  memory.WriteObject().Parameters() = arma::mat(parameters.memptr() + embedCount + searchCount + controllerCount + joinCount, writeCount, 1, false, true);
+  embed.Parameters() = arma::mat(parameters.memptr(), embedCount, 1, false, false);
```

Comment:
The problem with the current solution is, that we also have to reset the memory pointer for each layer, so I propose to use the ``NetworkInitialization`` class, which does all this for us.

Here is the updated ``ResetParameters`` method:


```c++
template<typename E, typename J, typename S, typename W, typename C>
void HAMUnit<E, J, S, W, C>::ResetParameters()
{
  embed.ResetParameters();
  search.ResetParameters();
  controller.ResetParameters();
  memory.ResetParameters();

  const size_t embedCount = embed.Parameters().n_elem;
  const size_t searchCount = search.Parameters().n_elem;
  const size_t controllerCount = controller.Parameters().n_elem;
  const size_t joinCount = memory.JoinObject().Parameters().n_elem;
  const size_t writeCount = memory.WriteObject().Parameters().n_elem;

  parameters = arma::mat(embedCount + searchCount + controllerCount +
      joinCount + writeCount, 1);

  NetworkInitialization<> networkInit;
  networkInit.Initialize(embed.Model(), parameters);
  networkInit.Initialize(search.Model(), parameters, embedCount);
  networkInit.Initialize(controller.Model(), parameters, embedCount +
      searchCount);
  networkInit.Initialize(memory.JoinObject().Model(), parameters, embedCount +
      searchCount + controllerCount);
  networkInit.Initialize(memory.WriteObject().Model(), parameters, embedCount +
      searchCount + controllerCount + joinCount);

  reset = true;
}
```

and here is the updated NetworkInitialization class: https://gist.github.com/zoq/a895915d2da1dc36770e2aea583b1147 it's bacially the same as used in #1066, just set a default ``InitializationRuleType``.

I've also passed the ``joiner`` and ``writer`` as reference in the ``TreeMemory`` class.